---
title: "Circtools"
author: "Alexey Uvarovskii"
date: '`r Sys.Date()`'
output:
  pdf_document:
    toc: yes
  rmarkdown::github_document:
    toc: yes
  rmarkdown::html_vignette:
    toc: yes
vignette: |
  %\VignetteIndexEntry{Vignette Title} %\VignetteEngine{knitr::rmarkdown} %\VignetteEncoding{UTF-8}
---


## Introduction 

The package aims to simplify analysis of the RNA-seq experiments for
circular RNA detection and quantification. 
Once a candidate list of possible circulirised RNA is provided 
(e.g. after analysis with splice-aware alignment tools like STAR and 
circRNA detection software like DCC or CIRI), it may be necessary 
to validate such transcripts by PCR.

We implemented a set of functions which could assist to design such 
validation experiments. A biologist might be interested to know

- which linear trabscripts are expressed and at which level
- which transcripts have common exons with the predicted circRNA
- how to design primers for the chosen linear and circular transcripts.



## A proposed workflow

1. Prepare the input data set:
    - the gene model: a GTF/GFF file, TxDb or EnsDb objects
    - transcripts counts 
    - splice junction coordinates for the circular candidates.
2. Generate sequences of the exons around the splice junctions.
3. Plot the gene model to see relation of linear and circular transcripts.
3. Design optimal primers for the circular and linear transcripts of interest.
4. Plot the gene model and the primers to validate the design.
  
  
## Transcript plot

One would like to know, how the predicted circRNA relate to the linear
transcripts.
In order to see, which transcripts are expressed and what is their gene model
(i.e. exon content), we implemented a plotting function.
It shows a block structure for every annotated transcript, a coordinate
range which is covered by the circRNA candidate and plots read counts for
the transcripts if they are provided.

## Annotation source 

Coordinates and types of genomic features must be known in advance, and
it can be provided in the form of `GTF/GFF` file or `TxDb`/`EnsDb` objects.
In the frame of the package we encourage one to use the Ensemble annotation
files or R packages, which can be downloaded from the [Ensembl site](http://www.ensembl.org) or installed via the Bioconductor ecosystem:
```{r eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("EnsDb.Hsapiens.v86")
```

In this example we are using the `ensembldb` package for annotation:
```{r}
suppressPackageStartupMessages(
  library("EnsDb.Hsapiens.v86")
)
db <- EnsDb.Hsapiens.v86
```

In the first step, the helper function create an object, which keep relations
between the annotated transcripts and coordinates for the predicted circRNAs.
We assume that a user have a `data.frame` or `GRanges` object with
splice junctions coordinates from which circRNAs are derived.
The input must contain the chromosome, start, end and the strand of the splice 
junction.

## Generate mock data

Here we create a splice junction GRanges:
```{r include=FALSE}
createCirc <- function(geneName, db){
  txs <- select(db, keys=geneName, keytype="GENENAME",columns="TXNAME")
  ex1 <- exonsBy(db, filter=TxidFilter(txs$TXNAME[1]))[[1]]
  ex2 <- exonsBy(db, filter=TxidFilter(txs$TXNAME[2]))[[1]]
  circCoords <- c(range(ex1[1:2]), range(ex2[1:2]))
  mcols(circCoords)$sjId <- paste0(seqnames(circCoords), ":",
                                     start(circCoords), "-", end(circCoords))
  circCoords
}

```

```{r}
geneName <- "BCL6"
circs <- createCirc(geneName, db)
circs
```

Let us simulate some numbers for read counts for transcripts of the BCL6
gene:

```{r include=FALSE}
makeCounts <- function(geneName, db) {
  ex <- exonsBy(db, by = "tx", filter = list(GenenameFilter(geneName)))
  txNames <- unique(names(ex))
  counts <- data.frame(id = names(ex),
  count = rnbinom(length(txNames), mu = 1000, size = 10))
  counts$count[sample(length(txNames), size = length(txNames) * .75)] <- 0
  counts
}

```

```{r}
counts <- makeCounts(geneName, db)
tail(counts)
```

## The workflow entry point

Using the splice junction table and annotation object, prepare the `CircData`
object:

```{r}
suppressPackageStartupMessages(
  library(circtools)
)
circData <- CircData(db, circs)
```

Plot the gene model, circular transcripts and read counts for the
BCL6 gene:
```{r}
bcl6EnsId <- circData$sjGeneIds
bcl6EnsId
```

```{r, fig.height=3, fig.width=6}
plotCirc(circGenes = bcl6EnsId,
         circData = circData,
         counts = counts)
```


Alternatively, the `plotTranscripts` function can be used for 
plotting from `data.frame` or `GRanges` objects provided by the user.
Since this function does not use any annotation information,
it plots all provided objects:

```{r, fig.height=3, fig.width=6}
ex <- ensembldb::exonsBy(db,  by = "tx",
                         filter = list(GenenameFilter(geneName)))
plotTranscripts(ex, counts = counts)
```

## Report sequences of the splice junction exons

Most probably, an experimentalist is interested to obtain the sequencies of
the splice junction exons. The sequencies will be used for the following
primer design to validate the discovered transcripts using the PCR.

To achieve it, besides the gene model annotation, 
one needs a fasta file or an R package with the corresponding
genome sequence.
 a BSgenome Bioconductor package can be used:
```{r}
suppressPackageStartupMessages(
  library(BSgenome.Hsapiens.NCBI.GRCh38))
bsg <- BSgenome.Hsapiens.NCBI.GRCh38
```
Several exons with the same start but different lengths can be included
in the annotation. By default, all described exons, which start or end at 
the position of the circular splice junction will be reported.
It is possible to include the shortes or the longest sequence by setting 
the `type` argument:
```{r cache=F}
# for all exons use
exSeqAll <- getExonSeqs(circData = circData, bsg = bsg, type = "all")
exShortesSeq <- getExonSeqs(circData = circData, bsg = bsg, type = "shortest")
exShortesSeq[['3:187734869-187737088']]
```

Experimentalists might be interested in obtaining a list with 
exon sequences and their coordinates for every circular splice junction.

Having the list of sequences from the `getExonSeqs` function and the `CircData`
object, let us create an HTML report to present it to the biologists:

```{r eval=FALSE}
reportCircs(exSeq = exShortesSeq, file = "report.html")
```

The resulting html file includes the information on the exon coordinates and
sequences for the following primer design.

![The HTML report with splice junction exon sequences.](img/report.png)

## Design and validate primers

To get *in silico* optimized primer sequences, one needs simply to 
invoke `designPrimers` function on the splice junction exons object:
```{r}
primers <- designPrimers(exSeq = exShortesSeq[1], db = db, bsg = bsg)
```

The result is a list with an item for every splice junction.
There are two records: `primers` and `products`.
Every item consists of a list of primers for possible circular transcripts:
if there several exons, which correspons to the same splice junction, 
all possible combinations of their pairs will be used for primer design.

The priducts are
```{r}
str(primers$products)
```
and the primers
```{r}
primers$primers$`3:187734869-187737088`
```

```{r}

plotCirc(circGenes = bcl6EnsId,
         circData = circData,
         counts = counts, 
         primers =(primers$primers$`3:187734869-187737088`))
```


## Session
```{r}
sessionInfo()
```

