---
title: "Transcript plot"
author: "Alexey Uvarovskii"
date: '`r Sys.Date()`'
output:
  pdf_document: default
  rmarkdown::html_vignette: default
vignette: |
  %\VignetteIndexEntry{Vignette Title} %\VignetteEngine{knitr::rmarkdown} %\VignetteEncoding{UTF-8}
---

## Introduction 

The package aims to simplify analysis of the RNA-seq experiments for
circular RNA detection and quantification. 
Once a candidate list of possible circulirised RNA is provided 
(e.g. after analysis with splice-aware alignment tools like STAR and 
circRNA detection software like DCC or CIRI), it may be necessary 
to validate such transcripts by PCR.

We implemented a set of functions which could assist to design such 
validation experiments. A biologist might be interested to know

- which linear trabscripts are expressed and at which level
- which transcripts have common exons with the predicted circRNA
- how to design primers for the chosen linear and circular transcripts.


```{r}
suppressPackageStartupMessages(
  library(circtools)
)
```

## A proposed workflow

1. Prepare the input data set:
    - the gene model: a GTF/GFF file, TxDb or EnsDb objects
    - transcripts counts 
    - splice junction coordinates for the circular candidates.
2. Generate sequences of the exons around the splice junctions.
3. Plot the gene model to see relation of linear and circular transcripts.
3. Design optimal primers for the circular and linear transcripts of interest.
4. Plot the gene model and the primers to validate the design.
  
## Report sequences of the splice junction exons
  
## Transcript plot

One would like to know, how the predicted circRNA relate to the linear
transcripts.
In order to see, which transcripts are expressed and what is their gene model
(i.e. exon content), we implemented a plotting function.
It shows a block structure for every annotated transcript, a coordinate
range which is covered by the circRNA candidate and plots read counts for
the transcripts if they are provided.

Coordinates and types of genomic features must be known in advance, and
it can be provided in the form of `GTF/GFF` file or `TxDb`/`EnsDb` objects.
In the frame of the package we encourage one to use the Ensemble annotation
files or R packages, which can be downloaded from the [Ensembl site](http://www.ensembl.org) or installed via the Bioconductor ecosystem:
```{r eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("EnsDb.Hsapiens.v86")
```

In this example we are using the `ensembldb` package
```{r}
library("EnsDb.Hsapiens.v86")
db <- EnsDb.Hsapiens.v86
```

In the first step, the helper function create an object, which keep relations
between the annotated transcripts and coordinates for the predicted circRNAs.
Here, we assume that a user have a `data.frame` or `GRanges` object with
splice junctions coordinates from which circRNAs are derived.
The input must contain the chromosome, start, end and the strand of the splice 
junction.

Let us simulate some numbers for read counts for transcripts of the BCL6
gene:

```{r}
makeCounts <- function(geneName, db) {
  ex <- exonsBy(db, by = "tx", filter = list(GenenameFilter(geneName)))
  txNames <- unique(names(ex))
  counts <- data.frame(id = names(ex),
  count = rnbinom(length(txNames), mu = 1000, size = 10))
  counts$count[sample(length(txNames), size = length(txNames) * .75)] <- 0
  counts
}
geneName <- "BCL6"
counts <- makeCounts(geneName, db)
```

```{r}

createCirc <- function(geneName, db){
  txs <- select(db, keys=geneName, keytype="GENENAME",columns="TXNAME")
  ex1 <- exonsBy(db, filter=TxidFilter(txs$TXNAME[1]))[[1]]
  ex2 <- exonsBy(db, filter=TxidFilter(txs$TXNAME[2]))[[1]]
  circCoords <- c(range(ex1[1:2]), range(ex2[1:2]))
  mcols(circCoords)$CIRCID <- paste0(seqnames(circCoords), ":",
                                     start(circCoords), "-", end(circCoords))
  circCoords
}
circs <- createCirc(geneName, db)
circData <- CircData(db, circs)
```
```{r}
plotCirc(circGenes = circData$sjGeneIds,circData = circData, counts = counts)
```


Alternatively, the `plotTranscripts` function can be used for 
plotting from `data.frame` or `GRanges` objects provided by the user.
Since this function does not use any annotation information,
it plots all provided objects:

```{r, fig.height=3, fig.width=6}
ex <- exonsBy(db, by = "tx", filter = list(GenenameFilter(geneName)))
plotTranscripts(ex, counts = counts)
```

## Design and validate primers

## Session
```{r}
sessionInfo()
```

